ambry-cloud/src/main/java/com/github/ambry/cloud/azure/AzureCloudDestination.java

```diff
@@ -151,13 +152,30 @@ public void downloadBlob(BlobId blobId, OutputStream outputStream) throws CloudS
   }
 
   @Override
-  public boolean deleteBlob(BlobId blobId, long deletionTime) throws CloudStorageException {
-    return updateBlobMetadata(blobId, CloudBlobMetadata.FIELD_DELETION_TIME, deletionTime);
+  public boolean deleteBlob(BlobId blobId, long deletionTime, short lifeVersion) throws CloudStorageException {
+    Map<String, Object> updateFields = new HashMap<>();
+    // TODO Frontend support needs to handle the special case of life version = MessageInfo.LIFE_VERSION_FROM_FRONTEND
+    updateFields.put(CloudBlobMetadata.FIELD_LIFE_VERSION, lifeVersion);
+    updateFields.put(CloudBlobMetadata.FIELD_DELETION_TIME, deletionTime);
+    return updateBlobMetadata(blobId, updateFields);
   }
 
   @Override
   public boolean updateBlobExpiration(BlobId blobId, long expirationTime) throws CloudStorageException {
-    return updateBlobMetadata(blobId, CloudBlobMetadata.FIELD_EXPIRATION_TIME, expirationTime);
+    return updateBlobMetadata(blobId,
+        Collections.singletonMap(CloudBlobMetadata.FIELD_EXPIRATION_TIME, expirationTime));
+  }
+
+  @Override
+  public short undeleteBlob(BlobId blobId, short lifeVersion) throws CloudStorageException {
+    Map<String, Object> updateFields = new HashMap<>();
+    updateFields.put(CloudBlobMetadata.FIELD_LIFE_VERSION, lifeVersion);
+    updateFields.put(CloudBlobMetadata.FIELD_DELETION_TIME, Utils.Infinite_Time);
+    updateBlobMetadata(blobId, updateFields);
+    // We either update lifeVersion or throw error. So this should work for now.
```

Comment:
in `updateBlobMetadata`, there are three cases possible
1. there is an exception during the update.
2. update is successful and we return true.
3. update returns false, because metadata is already updated to the asked values.

So, in effect, either the lifeversion gets updated, or an error is thrown.
Anyways, with the upcoming cloudblob store changes to support frontend request, I will handle this case to return the correct lifeVersion. So this is a temporary fix to enable replication, and unblock undelete testing, while I am working on cloud blob store related changes.